Add ssh transport to clone, push, pull, and fetch

andersonmills opened the pull request:
With this commit, ssh with PKI is added as a transport option for clone, push, pull, and fetch.

andersonmills commented:
@bartkamphorst There are a few questions I have about this implementation.

- This implementation doesn't look forward to using ssh with passwords, given that :username in the options is a trigger to use the UsernamePasswordCredentialsProvider. How would y'all like to specify the transport separate from username? Could only the :transport option be used? That is a breaking change for the current implementation. Optionally, :ssh_username could be specified for the ssh transport username for password use.
- I'm guessing the name RJGitTransportConfigCallback may not be the final name you want for that class.
- It looks like this could be added pretty quickly to the push and pull commands, too. Are y'all interested in that?

bartkamphorst commented:
@andersonmills Could the url of the remote (e.g., `repo.config['remote origin']['url']`) be a useful heuristic for determining transport protocol? Otherwise we could make http/https the default transport protocol and let people deviate from the default with `:transport_protocol = ":ssh"` (that way we wouldn't break compatibility).

I don't mind the long class name for `RJGitTransportConfigCallback`, but seeing how it is only used for ssh, perhaps RJGitSSHConfigCallback is more appropriate? Just a thought. @dometto ?

And yes, I'd really like to see ssh transport support for push and pull as well!

>  I'll make sure that not specifying a private_key_file just "works" and specifying one also works

That would be terrific. Thanks again for your efforts!

dometto commented:
+1 for determining transport protocol from the url, if the remote url 
contains the schema (URI.parse(remote).schema spares you the need to 
construct your own regex, I think). Otherwise I agree a separate 
:transport_protocol option would be cleaner than having both 
:ssh_username, :http_username, etc.

Yes, RJGitSSHConfigCallback, since the callback is specifically aimed at 
SSH.

All in all, great work on this and the other PRs @andersonmills!

andersonmills commented:
Glad to help. What I'm doing is being supported by Puppet Labs, so they get some credit, too. :)

The additions for ssh transport to work with a user's ssh configuration is now in. They can also use an encrypted ssh id file, too, though they'd have to deal with their own security issues around getting that passphrase.

I'll be working on the other suggestions and the testing of ssh, now.

andersonmills commented:
Looking at a style guide, https://github.com/bbatsov/ruby-style-guide#classes--modules, I'm told that the RJGitSSHConfigCallback class should go in its own file under /lib/git. This is the only internal class that I see. What would y'all prefer?

andersonmills commented:
URI scheme detection now added. This ends up being somewhat complicated, but usable.

ssh is used if URI is ssh or `:ssh` is specified as `option[:transport_protocol]` or `option[:private_key_file]` has a value.

UserPasswordCredentialsProvider is used if URI is https or http, or if `options[:username]` has a value. I left out the choice to specify `option[:transport_protocol]` as `:http` or `:https` because you must put that in the URI also, so that option doesn't have an effect.

options[:username] should be used appropriately in all cases, and a non-transport-specifying URI should work correctly, as long as it is correct (file or ssh).

andersonmills commented:
I'm definitely open to suggestions on testing this.

bartkamphorst commented on lib/git.rb:
```diff
@@ -127,6 +130,55 @@ def clone(remote, local, options = {})
+      end
+      transport_protocol = uri.scheme if uri
+
+      if ("ssh" == transport_protocol) || (options[:transport_protocol] == :ssh) || options[:private_key_file]
```
Question. In the scenario that `transport_protocol` cannot be derived from `remote` and no private key is specified, yet `:transport_protocol` is set to `:ssh`, what are the odds that the command will succeed?

andersonmills replied:
If the user has a .ssh/config file with a correct entry for the host of the remote in it, then it will succeed. This was my accidental first test case. :)

bartkamphorst commented on lib/git.rb:
```diff
@@ -127,6 +130,55 @@ def clone(remote, local, options = {})
+
+      if ("ssh" == transport_protocol) || (options[:transport_protocol] == :ssh) || options[:private_key_file]
+        command.set_transport_config_callback(RJGitSSHConfigCallback.new(options[:private_key_file], options[:private_key_passphrase]))
+      elsif (["http","https"].include? transport_protocol) || options[:username]
```
So `options[:username]` and `options[:password]` are never used for ssh?

andersonmills replied:
Not at the moment, as that was not our use case, however I believe it's a super quick change to the override of configure in the anonymous class inheriting from JschConfigSessionFactory.

I'll try to add that.

bartkamphorst commented:
> 'm definitely open to suggestions on testing this.

I'd like to see the rescue clause of `set_command_transport` tested. Also, I'd be interested to see/test the behavior when a bad `private_key_file` (such as `random_string`) is specified.

bartkamphorst commented:
> RJGitSSHConfigCallback class should go in its own file under lib/git

What about `lib/ssh` or `lib/callbacks`? @dometto ?

dometto commented:
There is also the `LocalRefWriter` class in `repo.rb` at the moment. I say we either follow the style guide strictly (so have a `lib/git` and `lib/repo`) or just make a generic folder for such small JGit implementations (something like `lib/interfaces` or `lib/classes`?). But maybe we should forget about that for this PR and restructure after this is merged.

andersonmills commented:
@bartkamphorst - I guess in the testing side of things, I'm just wondering if y'all are fine with unit tests of the methods, as integration tests are going to be hard to set up without a git server with ssh configured. Your other test suggestions are definitely planned.

@dometto I'm fine with leaving the internal class where it is and letting y'all do a bigger restructure of those things later.

bartkamphorst commented:
> integration tests are going to be hard to set up without a git server with ssh configured

Agreed. Unit tests are fine!

andersonmills commented:
Out of curiosity, did coveralls get turned off?

bartkamphorst commented:
No, still seems [operational](https://coveralls.io/builds/4975094).

andersonmills commented:
Weird. It didn't appear to run for the last commit for some reason.

andersonmills commented:
So, ssh on fetch is working now. push and pull are still in testing.

andersonmills commented:
Push is working. I'm having a problem with pull, but it may be my testing set up. Still working...

andersonmills commented:
Hunh. Without any changes, pull works, too. However, only the remote name according to `git remote` may be used. That's different than clone (only uri), fetch(uri or `git remote` name), and push(uri or `git remote` name).

andersonmills commented:
Okay, I'm happy with where this is. ssh is added to fetch, clone, push and pull. I functionally tested them all, with both keys and with username/password. There were two things that were messy. One was the different usage of remote for pull, but that's JGit's behavior. The second was that the key in known_hosts that was created for my when I ssh-ed into the test box was apparently incompatible with whatever JGit was using, so I had to generate an rsa key and put it in there. Then it worked fine.

Let me know if there are things you want me to change/fix up here.

bartkamphorst commented:
Terrific! The only thing I am not entirely satisfied with yet are the lines in `set_command_transport` where transport_protocol is checked. What about the following alternative?
```ruby
def self.set_command_transport(command, remote, options = {})
  uri = nil
  begin
  uri = URI.parse(remote) if remote
  rescue URI::InvalidURIError
  end
  transport_protocol = uri.scheme if uri
  transport_protocol ||= options[:transport_protocol]

  if transport_protocol.to_s == "ssh" || options[:private_key_file]
    command.set_transport_config_callback(RJGitSSHConfigCallback.new(options))
  elsif ["http","https"].include? transport_protocol && options[:username]
    command.set_credentials_provider(UsernamePasswordCredentialsProvider.new(options[:username], options[:password]))
  end
end
```

andersonmills commented:
Changed to your version of set_command_transport.

bartkamphorst commented:
Apologies, I should have mentioned that the proposal was untested.

andersonmills commented:
Hah. It turns out that your proposal works as well as mine did, but points out a difference in URI#scheme and how the git CLI git. I had accidentally not tested the following error conditions.

"git.server:gituser/repo.git" has a URI#scheme of "git.server", so won't work unless `options[:transport_protocol] is set to :ssh. (I only tested the positive.)

It also turns out that "git@gitserver.zzz:gituser/gitrepo.git" is an invalid URI, and also won't work unless `options[:transport_protocol] is set to :ssh. (Again, I only tested the positive.)

I'm trying to figure out what the right solution is here.

andersonmills commented:
I switched back to checking for transport_protocol or options[:transport_protocol] separately. I also added a check to see if URI#scheme matches one of the valid schemes for before it will set transport_protocol. That prevents a URI of "gitserver.zzz:gituser/gitrepo.git" from setting the transport_protocol to "gitserver.zzz".

The end result for that case is that the user will have to set either `option[:transport_protocol]` or `option[:private_key_file]` to use git CLI-like URIs with the ssh transport.

I also added a few test cases including the negatives of the positives above and that unknown URIs pass through without setting an ssh transport callback or a credential provider.

bartkamphorst commented:
Nice, I'm glad you caught you those cases!

bartkamphorst commented on lib/git.rb:
```diff
@@ -127,6 +133,64 @@ def clone(remote, local, options = {})
+
+      if (SSH_TRANSPORTS.include? transport_protocol.to_s) || (options[:transport_protocol] == :ssh) || options[:private_key_file]
+        command.set_transport_config_callback(RJGitSSHConfigCallback.new(options))
+      elsif (HTTP_TRANSPORTS.include? transport_protocol.to_s) || options[:username]
```
I think `&&` is now more appropriate than `||`.

andersonmills replied:
The II allows for the case where you have a username/password for a local repo.
Here is a test case you have for this:
https://github.com/repotag/rjgit/blob/master/spec/git_spec.rb#L270

bartkamphorst commented on lib/git.rb:
```diff
@@ -127,6 +133,64 @@ def clone(remote, local, options = {})
+        uri = URI.parse(remote) if remote
+      rescue URI::InvalidURIError
+      end
+      transport_protocol = uri.scheme if uri && (valid_transports.include? uri.scheme)
```
Stylistically I would prefer `if uri && valid_transports.include?(uri.scheme)`.

bartkamphorst commented on lib/git.rb:
```diff
@@ -127,6 +133,64 @@ def clone(remote, local, options = {})
+    end
+
+    def self.set_command_transport(command, remote, options = {})
+      valid_transports = SSH_TRANSPORTS + HTTP_TRANSPORTS + FILE_TRANSPORTS
```
Seems like a waste to assign this every time that `set_command_transport` is called. Let's make `valid_transports` into a constant of its own (`VALID_TRANSPORTS`; declared at the top).

andersonmills commented:
Given that local repos are tested with credentials, leaving the && as II makes more sense to me. (Otherwise it will break tests.)

I made the other changes you suggested, though.

And, I'm demoing the changes to my company tomorrow, apparently.

bartkamphorst commented:
>  Given that local repos are tested with credentials, NOT changing the II to && makes more sense to me. 

Good call; agreed.

> And, I'm demoing the changes to my company tomorrow, apparently.

That's exciting news. Curious to hear how it goes and what they think!

@dometto Anything you'd like to add before I merge?

dometto commented:
Sorry, I've been a bit busy. But this looks good, please do merge. @bartkamphorst: we should decide about where to put the callback and refwriter classes and make sure there are some docs for the new features.

@andersonmills, thanks again and good luck with the demo. :)

bartkamphorst commented:
:+1:

andersonmills commented:
@bartkamphorst @dometto - The demo went just fine. We were more telling people what our team had been up to than really showing them new things they would be using. This work is an integral part of pulling apart a piece of software into a library that new software from our team will be using, so it didn't have a broad effect. The one question I did get was about whether we had any metrics around how rjgit performs compared to CLI git, but we haven't run metrics on that yet.

bartkamphorst commented:
Thanks for letting us know! Glad the demo went over well. If you do happen to run any metrics, we'd be interested to know about them!

bartkamphorst commented:
@andersonmills FYI, we just released a new version of the gem that includes your contributions. Thanks again!

andersonmills commented:
@bartkamphorst Thanks for the heads up. I'll let our release engineering team know. 

Also, there are a couple of other features that aren't currently in RJGit, which are available in JGit, that we may come back to work on soon. We have a few other problems to solve first, but I'll jump in with issues when it makes sense.

bartkamphorst commented:
Sounds good!
